web/pdf_page_view.js

```diff
@@ -1240,4 +1277,229 @@ class PDFPageView extends PDFPageViewBase {
+    // visibleLength * (2 * overflowScale + 1).
+    // We default to adding a whole height/length of detail area on each side,
+    // but we can reduce it to make sure that we stay within the maxCanvasPixels
+    // limit.
+    const visiblePixels =
+      visibleWidth * visibleHeight * (window.devicePixelRatio || 1) ** 2;
+    const maxDetailToVisibleLinearRatio = Math.sqrt(
+      maxCanvasPixels / visiblePixels
+    );
+    const maxOverflowScale = (maxDetailToVisibleLinearRatio - 1) / 2;
+    let overflowScale = Math.min(1, maxOverflowScale);
+    if (overflowScale < 0) {
+      overflowScale = 0;
+      // In this case, we render a detail view that is exactly as big as the
+      // visible area, but we ignore the .maxCanvasPixels limit.
+      // TODO: We should probably instead give up and not render the detail view
+      // in this case. It's quite rare to hit it though, because usually
+      // .maxCanvasPixels will at least have enough pixels to cover the visible
+      // screen.
+    }
+
+    const overflowWidth = visibleWidth * overflowScale;
+    const overflowHeight = visibleHeight * overflowScale;
+
+    const minX = Math.max(0, visibleArea.minX - overflowWidth);
+    const maxX = Math.min(viewport.width, visibleArea.maxX + overflowWidth);
+    const minY = Math.max(0, visibleArea.minY - overflowHeight);
+    const maxY = Math.min(viewport.height, visibleArea.maxY + overflowHeight);
+    const width = maxX - minX;
+    const height = maxY - minY;
+
+    this.detailArea = { minX, minY, width, height };
+
+    this.reset({ keepCanvas: true });
+  }
+
+  async draw() {
+    const initialRenderingState = this.renderingState;
+    if (initialRenderingState !== RenderingStates.INITIAL) {
+      console.error("Must be in new state before drawing");
+      this.reset(); // Ensure that we reset all state to prevent issues.
+    }
+    const { div, pdfPage, viewport } = this.pageView;
+
+    if (!pdfPage) {
+      this.renderingState = RenderingStates.FINISHED;
+      throw new Error("pdfPage is not loaded");
+    }
+
+    this.renderingState = RenderingStates.RUNNING;
+
+    const canvasWrapper = this.pageView._ensureCanvasWrapper();
+
+    const { canvas, prevCanvas, ctx } = this._createCanvas(
+      newCanvas => {
+        // If there is already the background canvas, inject this new canvas
+        // after it. We cannot simply use .append because all canvases must
+        // be before the SVG elements used for drawings.
+        if (canvasWrapper.firstElementChild?.tagName === "CANVAS") {
+          canvasWrapper.firstElementChild.after(newCanvas);
+        } else {
+          canvasWrapper.prepend(newCanvas);
+        }
+      },
+      // If there is already the lower resolution canvas behind,
+      // we don't show the new one until when it's fully ready.
+      this.pageView.renderingState === RenderingStates.FINISHED ||
+        initialRenderingState === RenderingStates.FINISHED
```

Comment:
I might be missing something, but can we actually reach this spot if the initial RenderingState is FINISHED?